Registrar script first draft

cisaacstern opened the pull request:
This PR translates https://github.com/pangeo-forge/pangeo-forge-orchestrator/pull/55 into a standalone script that:

- Queries the API (based on a recipe run primary key) to figure out the correct repo / commit to check out
- Retrieves the recipe module code using PyGithub and caches it locally (AFAICT, this is required to import using `importlib.utils`)
- Builds the flow
- Registers the flow

It works as-is, providing the correct env variables are set, which are:
- a `GITHUB_TOKEN` (required by PyGitHub)
- a `PANGEO_FORGE_API_URL` (currently the only one with data in it is https://api-staging.pangeo-forge.org/)

Plus Prefect and storage details + creds. Which, with the above two variables already set, would make calling the script look something like

```
PROJECT_NAME="pangeo-forge-4967" \
GOOGLE_APPLICATION_CREDENTIALS="<path to json creds for pangeo-forge-4967 project>" \
STORAGE_NAME="pfcsb-bucket" \
PANGEO_FORGE_OSN_KEY="<osn key>" \
PANGEO_FORGE_OSN_SECRET="<osn secret>" \
BAKERY_IMAGE="pangeo/pangeo-forge-bakery-images:pangeonotebook-2021.07.17_prefect-0.14.22_pangeoforgerecipes-0.6.1" \
PREFECT_PROJECT="pangeo-forge" \
PREFECT__CLOUD__AUTH_TOKEN="<prefect auth token>" \
PREFECT__CLOUD__AGENT__LABELS="{}" \
python3 registrar.py 1 "test-registrar-script-0"
```

There's a lot of generalization left to do, but I thought I'd pause here to get some feedback on what I already have. If this looks good, next steps would be:

- [ ] Run the script in a pre-configured docker container that includes all of the software dependencies pre-installed, and takes env vars passed from the calling context
- [ ] Decide on the calling context. I'd initially thought FastAPI, but since the script is quite generalizable in its current form, I'm now wondering if GitHub repository dispatch (with credentials etc. stored as GitHub Secrets) might actually work after al.

cisaacstern commented:
(testing slack subscription for this repo)

rabernat commented on registrar/flow.py:
```diff
@@ -0,0 +1,104 @@
+        image=os.environ["BAKERY_IMAGE"],
+        labels=json.loads(os.environ["PREFECT__CLOUD__AGENT__LABELS"]),
+        cpu_request="1000m",
+        memory_request="3Gi",
```
Is this enough memory?

cisaacstern replied:
Hmm good question. This is copied directly from https://github.com/pangeo-forge/pangeo-forge-gcs-bakery/blob/2d164b4dacb0353d1df11ebf996493e5108bc7bf/test/recipes/oisst_recipe.py#L74

@sharkinsspatial, thoughts?

cisaacstern replied:
I started https://github.com/pangeo-forge/registrar/issues/2 to track this. My suggestion is that we punt on this for now, in the interest of getting the "steel thread" connected, but revisit it before launching the public service.

rabernat commented on registrar/flow.py:
```diff
@@ -0,0 +1,104 @@
+            "pod_template": make_pod_spec(
+                image=os.environ["BAKERY_IMAGE"],
+                labels={"flow": flow_name},
+                memory_limit="1Gi",
+                memory_request="500Mi",
```
I thought we guaranteed 4GB memory for the workers?

cisaacstern replied:
Right. I didn't change any settings from the test flow linked above in https://github.com/pangeo-forge/registrar/pull/1#discussion_r801325709.

Perhaps we want different settings for this real recipe run (as opposed to the linked test script).

cisaacstern replied:
xref https://github.com/pangeo-forge/registrar/pull/1#discussion_r802092088 & #2 ... I suggest we punt for now, but make sure to revisit soon.

rabernat commented on registrar/flow.py:
```diff
@@ -0,0 +1,104 @@
+        use_listings_cache=False,
+    )
+    target_base = "s3://Pangeo/pangeo-forge"
+    recipe.target = MetadataTarget(fs_osn, root_path=f"{target_base}/pfcsb-test/{flow_name}")
```
```suggestion
    recipe.target = FSSpecTarget(fs_osn, root_path=f"{target_base}/pfcsb-test/{flow_name}.zarr")
```

Also, we should be using the recipe name / id, as [specified in ARD 2](https://github.com/pangeo-forge/roadmap/blob/master/doc/adr/0002-use-meta-yaml-to-track-feedstock-metadata.md#top-level-data) as the name. See also https://github.com/pangeo-forge/roadmap/blob/master/doc/adr/0003-standardize-storage-target-layout.md

cisaacstern replied:
> Also, we should be using the recipe name / id, as [specified in ARD 2](https://github.com/pangeo-forge/roadmap/blob/master/doc/adr/0002-use-meta-yaml-to-track-feedstock-metadata.md#top-level-data) as the name.

✔️ https://github.com/pangeo-forge/registrar/pull/1/commits/63a295487698cfb7b22e6c5174ce83748e5ed1b1

> See also https://github.com/pangeo-forge/roadmap/blob/master/doc/adr/0003-standardize-storage-target-layout.md

This ADR says we should include `feedstock_name` in the path. Currently, our database doesn't store this information exactly; instead, we have a `Feedstock.spec`, which might be, e.g., [`"pangeo-forge/noaa-oisst-avhrr-only-feedstock"`](https://api-staging.pangeo-forge.org/feedstocks/1). Including this whole string seems redundant (because "pangeo-forge" is captured elsewhere in the path)... I wonder if we maybe don't need to encode this information in the paths anymore, now that we can link datasets back to feedstocks via the database?

rabernat commented on registrar/registrar.py:
```diff
@@ -0,0 +1,59 @@
+    # import recipe
+    spec = importlib.util.spec_from_file_location("cached_recipe", cache_path)
+    imported_recipe_module = importlib.util.module_from_spec(spec)
+    spec.loader.exec_module(imported_recipe_module)
```
Just noting that this is where the danger lies if we consider the recipe code as "untrusted".

rabernat commented on registrar/registrar.py:
```diff
@@ -0,0 +1,59 @@
+if __name__ == "__main__":
+    args = sys.argv
+    recipe_run_primary_key = args[1]
+    flow_name = args[2]
```
What do you imagine this will be?

cisaacstern replied:
Good question. In an f-string which [combines the `recipe_run.started_at` timestamp with the `recipe_id`](https://github.com/pangeo-forge/registrar/pull/1/commits/65103faded35c43047049eb6e401a28b01c776a2#diff-db017ffd3c57141f7abbdf553a540518334b277df4ff9b917ea8465071107b0eR61). For the OISST feedstock, this gives a flow name such as `20220204T212105Z-pfcsb-test/noaa-oisst-pruned`.

Admittedly the forward slash in the `recipe_id` is un-Pythonic (maybe we disallowed these in an ADR somewhere?), but that's an artifact of the sample data I loaded into the staging database. Aside from that, what do you think of this for a generated flow name?

rabernat replied:
What is the flow name used for?

cisaacstern replied:
Prefect uses this name as the identifier for the flow on the Prefect Cloud dashboard.

rabernat replied:
Then I agree it should be set automatically. I would probably go with `{feedstock}-{commit}-{recipe_name}-{date}`

On a related note, the dataset name in the storage target should follow https://github.com/pangeo-forge/roadmap/blob/master/doc/adr/0003-standardize-storage-target-layout.md

rabernat replied:
Do the flow names have any requirement to be unique? If not, just use the recipe name

cisaacstern replied:
> Do the flow names have any requirement to be unique? If not, just use the recipe name

Yes, flow names need to be unique. Registering non-unique names results in the preexisting flow of the same name getting a version increment (rather than a new flow being created).

I think a workable solution here is what I've done in https://github.com/pangeo-forge/registrar/pull/1/commits/f277b5113cfe96ddad607511be4c8fbfebde2b34, where we assign

```python
flow.name = f"{recipe_run.recipe_id}-{recipe_run.head_sha}"[:63]
```
which gives us readability from `recipe_id` plus a some level of uniqueness assurance from a truncated `head_sha`. The likelihood that two feedstocks will define the same `recipe_id` (which I believe only has to be locally unique within the feedstock namespace), _and_ that the first dozen plus characters of the `head_sha` would also match, seems vanishingly small.

(Another thing I learned just now is that flow names can't have slashes in them, so I swapped a `-` for the previous `/` in our staging database test data.)

So this format gives us, e.g. `pfcsb-test-noaa-oisst-pruned-cd25e2914f699d4580b6a0afe8b9aab0b7` for recipe run 1 from the staging database.

cisaacstern replied:
> On a related note, the dataset name in the storage target should follow https://github.com/pangeo-forge/roadmap/blob/master/doc/adr/0003-standardize-storage-target-layout.md

https://github.com/pangeo-forge/registrar/commit/f277b5113cfe96ddad607511be4c8fbfebde2b34 does this.

Our typical OSN `target_base` string would be `s3://Pangeo/pangeo-forge`. Interestingly, we can actually fulfill the ADR's spec of

```
{prefix}/pangeo-forge/{feedstock_name}/ ... etc
```

with just

```python
f"s3://Pangeo/{recipe_run.feedstock.spec}/ ... etc
```

because our GitHub `feedstock.spec`s begin with the GitHub org name; i.e. are of the format `pangeo-forge/<feedstock_name>`

rabernat reviewed (commented):
Looking good.

rabernat commented:
It seems like the following config

```
STORAGE_NAME="pfcsb-bucket" \
PANGEO_FORGE_OSN_KEY="<osn key>" \
PANGEO_FORGE_OSN_SECRET="<osn secret>" \
```

could be injected into the bakery environment directly, rather than bundled into the flow.

rabernat commented:
To me, these things are part of the bakery configuration, not the recipe itself. So they should not live the with recipe.

To include them in the bakery environment by default, we would add them to a deployment-specific configuration associated with https://github.com/pangeo-forge/pangeo-forge-gcs-bakery. Maybe in the .env file.

cisaacstern commented:
> To me, these things are part of the bakery configuration, not the recipe itself. So they should not live the with recipe.

I've been assuming that these need to be bundled in the flow, because to generate the flow we compile the recipe with `recipe.to_prefect()`, and the storage targets need to be attached to the recipe object before that happens? (And to instantiate the storage objects, we need these variables.)

rabernat commented:
Right, the way things currently work, I think that's correct. So no action to take on this right now.

Going forward, the approach described here - https://github.com/pangeo-forge/roadmap/pull/36 - would allow us to dynamically load secrets at runtime.

cisaacstern commented:
I needed to push https://github.com/pangeo-forge/registrar/pull/1/commits/fa6bc0d8960f4917d3b08501d3a22bfa4e072023 to get `copy_pruned` to work for the OISST recipe without raising

```python
---------------------------------------------------------------------------
Task 'store_chunk[0]': Exception encountered during task execution!
Traceback (most recent call last):
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/prefect/engine/task_runner.py", line 861, in get_task_run_state
    value = prefect.utilities.executors.run_task_with_timeout(
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/prefect/utilities/executors.py", line 323, in run_task_with_timeout
    return task.run(*args, **kwargs)  # type: ignore
  File "/Users/charlesstern/Dropbox/pangeo/registrar/registrar/flow.py", line 38, in wrapper
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/pangeo_forge_recipes/recipes/xarray_zarr.py", line 499, in store_chunk
    write_region, conflicts = region_and_conflicts_for_chunk(config, chunk_key)
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/pangeo_forge_recipes/recipes/xarray_zarr.py", line 232, in region_and_conflicts_for_chunk
    target_grid = ChunkGrid.from_uniform_grid(
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/pangeo_forge_recipes/chunk_grid.py", line 45, in from_uniform_grid
    assert (
AssertionError: invalid chunksize 20 and dimsize 2
```
I'm surprised that we didn't hit this earlier, but it seems like an actual bug (rather than user error). Basically, the un-pruned recipe defines `inputs_per_chunk=20` but the pruned recipe only has a total of 2 input files, and so fails when `region_and_conflicts_per_chunk` is called within `store_chunk`.

cisaacstern commented:
@rabernat, I believe I've addressed all of the points you raised:

- storage target paths should now match ADR spec
- flow run names are generated automatically with readable yet unique identifiers
- memory questions are tracked in #2

In the interest of moving fast for OSM prep, I propose we leave the patch described in https://github.com/pangeo-forge/registrar/pull/1#issuecomment-1033107536 as-in, and correct once we can find a more robust upstream fix for https://github.com/pangeo-forge/pangeo-forge-recipes/issues/279

IMO this is ready to merge, and we can move onto 

- [ ] dockerizing
- [ ] trying to call it with repository dispatch (+ GitHub Secrets) or similar

Anything I've overlooked or that you'd prefer to address further now?

cisaacstern commented:
Merging now but happy to revisit any of this through future issues/PRs!
